### PR TITLE
[Merged by Bors] - Fix most data races

### LINKED ISSUE #2226: Add -race flag to tests and fix data races
## Description
Neither unit tests, nor system tests use the `-race` flag. Using it would enable Go's race detector and let us find some existing data races easily. As a result of a read/write operation is undefined in case of a data race, they might cause some unexpected behavior.

I ran unit tests with the flag, and there are many races in multiple packages.

This issue appears in commit hash: 
`f790a44`

Related files (optionally with line numbers): 
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L137
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L142
https://github.com/spacemeshos/go-spacemesh/blob/develop/Makefile#L142
https://github.com/spacemeshos/go-spacemesh/blob/develop/Dockerfile#L31

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation

WIP on #2226

Depends on https://github.com/spacemeshos/post/pull/45

Some data races still exist and are triggered by logs that appear after tests have finished their execution. This still doesn't allow to enable the `-race` flag. However, most issues in the non-test code seem fixed. 

## Changes

- Fix most data races

## Test Plan

- Unit tests
- System tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: cmd/base.go
Comment (dshulyak) [inline cmd/base.go]: i don't like introducing global mutex. it was a mistake to introduce global context and cancel function in the first place, maybe it is possible to refactor code to avoid global context? or it is too many changes?

File: hare/algorithm.go
Comment (dshulyak) [inline hare/algorithm.go:81]: it is better to understand when k can be updated and read concurrently. the race here might be the symptom of the real issue

File: hare/algorithm.go
Comment (dshulyak) [inline hare/algorithm.go:898]: also not clear how the eligibility count can be read and updated concurrently

File: p2p/connectionpool/connectionpool.go
Comment (dshulyak) [inline p2p/connectionpool/connectionpool.go]: what if Wait completes and we call Add(1) afterwards? i think this is the bug and we need to check if connection pool is closed before Add(1)

File: tortoise/verifying_tortoise_test.go
Comment (dshulyak) [inline tortoise/verifying_tortoise_test.go]: logtest.New(t)

**Review by dshulyak — COMMENTED** [review]
there are several places where race is probably a symptom of the bug. for example if global context is overwritten and we call cancel on older context. in hare there are several suspicious places as well. i think it is better to understand why it can be called concurrently and reduce usage of shared memory, and only if the this is not possible then wrap everything in mutex.

otherwise we will simply hide the issues without addressing them

File: cmd/base.go
Comment (nkryuchkov) [reply cmd/base.go]: not too many changes, fixed

File: hare/algorithm.go
Comment (nkryuchkov) [reply hare/algorithm.go:91]: IIRC, k was concurrently updated in the test, therefore data races occurred

Comment (nkryuchkov) [reply hare/algorithm.go:81]: 1. 
```
Write at 0x00c000120318 by goroutine 25:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).addToK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:927 +0x84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).advanceToNextRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:545 +0x58
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:337 +0xf66
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:253 +0x58

Previous read at 0x00c000120318 by goroutine 26:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).getK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:917 +0x479
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).sendMessage()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:503 +0x4e0
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:300 +0x284

```

2. 
```
Write at 0x00c000266318 by goroutine 170:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).addToK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:927 +0x84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).advanceToNextRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:545 +0x58
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:337 +0xf66
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:253 +0x58

Previous read at 0x00c000266318 by goroutine 173:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).getK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:917 +0x33e
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:866 +0x3a5
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:292 +0x85
```

File: hare/algorithm.go
Comment (nkryuchkov) [reply hare/algorithm.go:898]:
```
Write at 0x00c00019e8ea by goroutine 174:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).currentRole()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:903 +0x5a8
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:864 +0x325
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).beginStatusRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:559 +0x2f7
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).onRoundBegin()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:694 +0x8d
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:341 +0xf84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:254 +0x58

Previous write at 0x00c00019e8ea by goroutine 193:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).currentRole()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:903 +0x5a8
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:864 +0x325
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:293 +0x85
```

File: hare/algorithm.go
Comment (nkryuchkov) [reply hare/algorithm.go:81]: @dshulyak 

It looks like the race detector also detects unsynchronized access at different times when there's actually no data race but there's a potential of a data race when timing differs. E. g., I increased the timeout in `TestConsensusProcess_TerminationLimit` to 10 seconds, and here's how logs look like:
```
==================
WARNING: DATA RACE
Write at 0x00c0002ce018 by goroutine 19:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).addToK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:939 +0x84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).advanceToNextRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:558 +0x58
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:344 +0x10d0
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:253 +0x58

Previous read at 0x00c0002ce018 by goroutine 20:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).getK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:929 +0x479
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).sendMessage()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:516 +0x4e0
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:304 +0x325
[...]
==================
sendMessage started at 2021-10-07 00:22:26.019217 +0300 MSK m=+0.073492008
[...]
sendMessage finished at 2021-10-07 00:22:26.019383 +0300 MSK m=+0.073658050
advanceToNextRound started at 2021-10-07 00:22:36.01865 +0300 MSK m=+10.072839518
advanceToNextRound finished at 2021-10-07 00:22:36.019665 +0300 MSK m=+10.073854166
```

So, here `advanceToNextRound` starts 10 seconds after `sendMessage` finishes, but the race detector still throws an error. I also tried a 1-minute timeout, the results are the same.

As I understand, the `ThreadSanitizer` library which is used as the Go race detector, logs every `atomic` operation by calling `__tsan_go_atomic32_load` (https://github.com/golang/go/blob/master/src/runtime/race_amd64.s#L200). Regarding mutexes, the compiler calls `__tsan_acquire` (https://github.com/golang/go/blob/master/src/sync/mutex.go#L76, https://github.com/golang/go/blob/master/src/runtime/race.go#L520) which establishes some relationships of what happens before what (https://github.com/golang/go/blob/master/src/runtime/race.go#L30) which ensures the lack of data race. So, what's not explicitly reported to TSAN is considered a data race.

I think we should just make the race detector happy and leave the access synchronized. If we need to ensure that the operations aren't done simultaneously, we should consider using other options

File: hare/algorithm.go
Comment (nkryuchkov) [reply hare/algorithm.go:898]: https://github.com/spacemeshos/go-spacemesh/pull/2608#discussion_r723713505

File: p2p/connectionpool/connectionpool.go
Comment (nkryuchkov) [reply p2p/connectionpool/connectionpool.go]: @dshulyak Could you please check if https://github.com/spacemeshos/go-spacemesh/pull/2608/commits/355c210289d00403834bc736a7239c0680b9be43 is what you meant?

**Comment by nkryuchkov** [comment]
Tests should pass after https://github.com/spacemeshos/post/pull/45 is merged

File: activation/activation.go
Comment (countvonzero) [inline activation/activation.go]: is there any particular reason to use RWMutex instead of Mutex? 

in my view, this is one form of premature optimization. RWMutex is suitable when there are more reads than writes. so in case where we only write, we should simply use Mutex for the following reasons:
- there is a slight performance penalty, but i can ignore that
- i have seen in huge codebase, debugging become more difficult with different kinds of lock and multiple lock holders.

i opt for simplicity for now and only use RWMutex when there are indeed a lot more readers than writers

File: activation/post.go
Comment (countvonzero) [inline activation/post.go]: rm

File: activation/post.go
Comment (countvonzero) [inline activation/post.go]: rm

File: activation/post.go
Comment (countvonzero) [inline activation/post.go:130]: does it make more sense to just get the lock once for this method, instead of 4 times, especially if this method is called frequently?

File: activation/post.go
Comment (countvonzero) [inline activation/post.go:227]: i think it make more sense to obtain the lock once for this method.
for simplicity, i'll even venture to use just 1 lock for everything in this file.

File: hare/consensus_test.go
Comment (countvonzero) [inline hare/consensus_test.go]: do you know why this requires extra time?

hare tests already take too long as is, now we are adding 10 seconds in many more tests?

**Review by countvonzero — COMMENTED** [review]
this is one tough PR to create. kudos to your effort.

enabling --race for tests is very important. i wholeheartedly support this effort.
however, i feel this blanketed approach with mutex is excessive. there are other tools such as making immutable struct (lower-case struct members+data hiding) or the atomic package.

i guess this is a good first step to force --race on all tests. but the code is unnecessarily bloated IMO.

**Comment by nkryuchkov** [comment]
@countvonzero 

Generally, I agree with your point of view. Most data races were fixed by using a mutex or atomics and their usage may look excessive in some places. However, this approach works, it gets rid of data races and is relatively easy to test. 

Deeper modifications would look better but they require much more time spent on the PR, we will need to resolve new merge conflicts constantly and will be having data races in `develop` until the PR is merged. Also, deeper modifications should be tested much better than just adding mutexes and atomics. A lot of packages are modified, so we'll need to test all parts of the projects, which is very risky and time-consuming.

In my opinion, it would be better for us to have smaller and shorter-living PRs with the least scope of changes, and to use an iterative approach, and to merge this PR as it is (after similar PR to PoET is merged). It's already better than what we have now in `develop`. We can continue making improvements in the next PRs.

File: activation/activation.go
Comment (nkryuchkov) [reply activation/activation.go]: You're right, `sync.RWMutex` is a premature optimization here, I'll change it to `sync.Mutex`

File: activation/post.go
Comment (nkryuchkov) [reply activation/post.go]: thanks, I forgot to remove it

File: activation/post.go
Comment (nkryuchkov) [reply activation/post.go:130]: I tried to reduce the number of mutex locks, but tests hang then. I tried to fix hanging but was able to do that only by reverting the number of locks. I think we need to leave it as is now and make deeper changes to the package and tests in the future

File: activation/post.go
Comment (nkryuchkov) [reply activation/post.go:227]: https://github.com/spacemeshos/go-spacemesh/pull/2608#discussion_r727415921

**Comment by countvonzero** [comment]
> @countvonzero
> 
> Generally, I agree with your point of view. Most data races were fixed by using a mutex or atomics and their usage may look excessive in some places. However, this approach works, it gets rid of data races and is relatively easy to test.
> 
> Deeper modifications would look better but they require much more time spent on the PR, we will need to resolve new merge conflicts constantly and will be having data races in `develop` until the PR is merged. Also, deeper modifications should be tested much better than just adding mutexes and atomics. A lot of packages are modified, so we'll need to test all parts of the projects, which is very risky and time-consuming.
> 
> In my opinion, it would be better for us to have smaller and shorter-living PRs with the least scope of changes, and to use an iterative approach, and to merge this PR as it is (after similar PR to PoET is merged). It's already better than what we have now in `develop`. We can continue making improvements in the next PRs.

i agree with all your stmts with the exception of the iterative approach. IMO, the iterative approach should be breaking down this issue into multiple packages and change them one at a time, instead of a blanketed change like this, which suffers exactly what you pointed out, never-ending merges.

what i am going to say is a little brutal. please pardon my lack of diplomacy.

in some way i feel huge changes like this is effectively putting a knife on others' throat and say "hey, i already done this much work, you need to accept it in order to move forward". i strongly discourage big changes like this going forward.

on that note, i'll approve and let's move on.

**Review by countvonzero — APPROVED** [review]

**Comment by nkryuchkov** [comment]
bors try

**Comment by nkryuchkov** [comment]
bors merge

**Pull request closed without merging**
